CommonTools/RecoUtils/plugins/LeptonFSRProducer.cc

```diff
@@ -0,0 +1,284 @@
+
+    descriptions.addWithDefaultLabel(desc);
+  }
+  ~LeptonFSRProducer() override {}
+
+private:
+  void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+
+  double computeRelativeIsolation(const pat::PackedCandidate& photon,
+                                  const pat::PackedCandidateCollection& pfcands,
+                                  const double& isoConeMax,
+                                  const double& isoConeMin) const;
+
+  bool electronFootprintVeto(pat::PackedCandidateRef& pfcandRef,
+                             edm::Handle<pat::ElectronCollection> electronsForVeto) const;
+
+  // ----------member data ---------------------------
+  const edm::EDGetTokenT<pat::PackedCandidateCollection> pfcands_;
+  const edm::EDGetTokenT<pat::ElectronCollection> electronsForVeto_;
+  const edm::EDGetTokenT<edm::View<reco::Muon>> muons_;
+  const edm::EDGetTokenT<edm::View<reco::GsfElectron>> electrons_;
+  float ptCutMu;
+  float etaCutMu;
+  float ptCutE;
+  float etaCutE;
+  float photonPtCut;
+  float drEtCut;
+  float isoCut;
+};
+
+void LeptonFSRProducer::produce(edm::StreamID streamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  using namespace std;
+
+  edm::Handle<pat::PackedCandidateCollection> pfcands;
+  iEvent.getByToken(pfcands_, pfcands);
+  edm::Handle<edm::View<reco::Muon>> muons;
+  iEvent.getByToken(muons_, muons);
+  edm::Handle<edm::View<reco::GsfElectron>> electrons;
+  iEvent.getByToken(electrons_, electrons);
+  edm::Handle<pat::ElectronCollection> electronsForVeto;
+  iEvent.getByToken(electronsForVeto_, electronsForVeto);
+
+  // The output collection of FSR photons
+  auto fsrPhotons = std::make_unique<std::vector<pat::GenericParticle>>();
+
+  std::vector<int> muPhotonIdxs(muons->size(), -1);
+  std::vector<double> muPhotonDRET2(muons->size(), 1e9);
+
+  std::vector<int> elePhotonIdxs(electrons->size(), -1);
+  std::vector<double> elePhotonDRET2(electrons->size(), 1e9);
+
+  //----------------------
+  // Loop on photon candidates
+  //----------------------
+
+  for (auto pc = pfcands->begin(); pc != pfcands->end(); pc++) {
+    // consider only photons, with pT and eta cuts
+    if (abs(pc->pdgId()) != 22 || pc->pt() < photonPtCut || abs(pc->eta()) > 2.5)
+      continue;
+
+    //------------------------------------------------------
+    // Get the closest lepton
+    //------------------------------------------------------
+    double dRMin(0.5);
+    int closestMu = -1;
+    int closestEle = -1;
+    float photon_relIso03 = 1e9;  // computed only if necessary
+    bool skipPhoton = false;
+
+    for (auto muon = muons->begin(); muon != muons->end(); ++muon) {
+      if (muon->pt() < ptCutMu || std::abs(muon->eta()) > etaCutMu)
+        continue;
+
+      int muonIdx = muon - muons->begin();
+      double dR = deltaR(muon->eta(), muon->phi(), pc->eta(), pc->phi());
+      if (dR < dRMin && dR > 0.0001 && dR / pc->pt() / pc->pt() < drEtCut) {
```

Comment:
Event though also this one could get optimized without loosing readability, e.g..:
```suggestion
      double dR = deltaR2(muon->eta(), muon->phi(), pc->eta(), pc->phi());
      if (dR < dRMin * dRMin && dR > 0.0001 * 0.0001) {
          dR = std::sqrt(dR);
          if (dR < drEtCut * pc->pt() * pc->pt()) {
```
etc...

Reply:
> Yes but this looks like a very tiny optimization, given that the number of candidates for FSR is small and the sqrt is done anyhow for those passing the dR cut (unless we also change the drEtCut). To me it seems to bring more confusion than speedup... but if you insist...

No, for this one I don't insist...